### DISCUSSION TITLE
spring_layout - negative weights

### DISCUSSION THREAD

**Discussion by horsto** [networkx/networkx#4811] [Q&A] [answered]
I have a naive question about the `weights` parameter for a [spring loaded model](https://networkx.org/documentation/stable/reference/generated/networkx.drawing.layout.spring_layout.html). I am constructing a `MultiGraph` object and I am adding edges with weights that center around 0. But what is actually a good range here and how does it affect the outcome of the simulation? 

I am getting the desired result when adding weights like that, i.e. I want negative values to have a repelling and positive values to have an attracting effect on connected nodes). But I am lacking intuition about what is actually correct here (or whether it matters at all).

**Comment by dschult** [comment, marked as answer]
The `spring_layout` treats `weights` as the spring constant of the springs (attractive force -- bigger brings the nodes closer).
This could probably be stated more clearly in the docstring.  But you seem to have got the meaning... negative values lead to a repelling force (in addition to the node repelling force) and positive values lead to an attracting force.  The method relies on a balance between node-based repelling forces and edge-based attracting forces. So centering the edge weights on zero isn't necessary. The center of the edge weights can be thought of as setting the balance between node-based repulsion and edge-based attraction. then the variation in edge weights causes relatively stronger or weaker attraction.  If you can find a better way to make this clear in the docs I would appreciate it. :}

**Reply by horsto** [reply]
Thanks a lot - your explanation helped to clarify!

Am I correct in assuming that the absolute amount of connections (edges) between nodes does not really matter if the (average) weight stays the same? Let's say I have 4 nodes (A,B,C,D) and the edges between C and D are under-sampled compared to all others. However, I am assuming that the distribution of weights between C and D are still representative of the population. Do I have to sub-sample all other edges to the number of edges between C and D or does this not matter? 

Also, do you have any practical tips on analysing / quantifying the outcome of the spring simulation? I see what I want to see (2 clusters in my 4 nodes A,B,C,D), but I am not sure what's the right way to show reliability? My naive approach was to use the positions of nodes at the end of the simulation and throw "standard" clustering methods on them (kMeans or similar) and compare clustering outcome to a distribution of simulations where weights were permuted / shuffled before building the graph. Does this make sense?

**Reply by dschult** [reply]
I'm not sure what the best way is to verify experimentally what the outcome of the spring layout simulation is. I think that probably depends on what kinds of questions you are asking about the layout process.  I have a similar response to your question about sub-sampling the edges. The edges are not created randomly in this algorithm. So, how the sub-sampling should play out depends on your process for generating the edge weights.
